### ISSUE TITLE
第7章 Transformer向け dataloader (torchtext 0.16.0対応)

### ISSUE THREAD

**Issue by pekujapan** [YutaroOgawa/pytorch_advanced#222]
せっかくいい本なので、もったいないと思い、どうにかして、現状のtorchtext用にデータローダ
作れないかトライしてましたが、なんとなく形になったので、以下示します。
まだ、学習させてないので、どうなるかわかりませんが、出力とlossは数値がでたようなので投稿しました。

 第7章 自然言語処理による感情分析（Transformer） 7_nlp_sentiment_transformer\utils\dataloader.py
import glob
import os
import io
import string
import re
import random
import spacy
import code
import torchtext
from torchtext.vocab import Vectors

def get_IMDb_DataLoaders_and_TEXT(max_length=256, batch_size=24,torchtext_version='0.11.0'):
    #IMDbのDataLoaderとTEXTオブジェクトを取得する。 

    writeenable=False
    # 訓練データのtsvファイルを作成します
    if writeenable:
       f = open('./data/IMDb_train.tsv', 'w')

       path = './data/aclImdb/train/pos/'
       for fname in glob.glob(os.path.join(path, '*.txt')):
           with io.open(fname, 'r', encoding="utf-8") as ff:
               text = ff.readline()
               # タブがあれば消しておきます
               text = re.sub('[\x80-\xff]', " ",text)
               text = re.sub('[\u0100-\uf0c0]', " ",text)
               text = text.replace('\t', " ")
               text = text+'\t'+'1'+'\n'
               f.write(text)

       path = './data/aclImdb/train/neg/'
       for fname in glob.glob(os.path.join(path, '*.txt')):
           with io.open(fname, 'r', encoding="utf-8") as ff:
               text = ff.readline()
               # タブがあれば消しておきます
               text = re.sub('[\x80-\xff]', " ",text)
               text = re.sub('[\u0100-\uf0c0]', " ",text)
               text = text.replace('\t', " ")
               text = text+'\t'+'0'+'\n'
               f.write(text)

       f.close()

       # テストデータの作成
       f = open('./data/IMDb_test.tsv', 'w')
       path = './data/aclImdb/test/pos/'
       for fname in glob.glob(os.path.join(path, '*.txt')):
           with io.open(fname, 'r', encoding="utf-8") as ff:
               text = ff.readline()
               # タブがあれば消しておきます
               text = text.decode(errors='ignore')
               text = re.sub('[\x80-\xff]', " ",text)
               text = re.sub('[\u0100-\uf0c0]', " ",text)
               text = text.replace('\t', " ")
               text = text+'\t'+'1'+'\n'
               f.write(text)

       path = './data/aclImdb/test/neg/'
       for fname in glob.glob(os.path.join(path, '*.txt')):
           with io.open(fname, 'r', encoding="utf-8") as ff:
               text = ff.readline()
               # タブがあれば消しておきます
               text = text.decode(errors='ignore')
               text = re.sub('[\x80-\xff]', " ",text)
               text = re.sub('[\u0100-\uf0c0]', " ",text)
               text = text.replace('\t', " ")
               text = text+'\t'+'0'+'\n'
               f.write(text)

       f.close()

    def preprocessing_text(text):
        # 改行コードを消去
        text = re.sub('<br />', '', text)
        # カンマ、ピリオド以外の記号をスペースに置換
        for p in string.punctuation:
            if (p == ".") or (p == ","):
                continue
            else:
                text = text.replace(p, " ")
        # ピリオドなどの前後にはスペースを入れておく
        text = text.replace(".", " . ")
        text = text.replace(",", " , ")
        return text

    # 分かち書き（今回はデータが英語で、簡易的にスペースで区切る）
    def tokenizer_punctuation(text):
        return text.strip().split()

    # 前処理と分かち書きをまとめた関数を定義
    def tokenizer_with_preprocessing(text):
        text = preprocessing_text(text)
        ret = tokenizer_punctuation(text)
        return ret

    if(torchtext_version=='0.11.0'):
       # データを読み込んだときに、読み込んだ内容に対して行う処理を定義します
       # max_length
       TEXT = torchtext.legacy.data.Field(sequential=True, tokenize=tokenizer_with_preprocessing, use_vocab=True,
                                   lower=True, include_lengths=True, batch_first=True, fix_length=max_length, init_token="<cls>", eos_token="<eos>")
       LABEL = torchtext.legacy.data.Field(sequential=False, use_vocab=False)

       # フォルダ「data」から各tsvファイルを読み込みます
       train_val_ds, test_ds = torchtext.legacy.data.TabularDataset.splits(
           path='./data/', train='IMDb_train.tsv',
           test='IMDb_test.tsv', format='tsv',
           fields=[('Text', TEXT), ('Label', LABEL)])

       # torchtext.data.Datasetのsplit関数で訓練データとvalidationデータを分ける
       train_ds, val_ds = train_val_ds.split(split_ratio=0.8, random_state=random.seed(1234))

       # torchtextで単語ベクトルとして英語学習済みモデルを読み込みます
       english_fasttext_vectors = Vectors(name='data/wiki-news-300d-1M.vec')

       # ベクトル化したバージョンのボキャブラリーを作成します
       TEXT.build_vocab(train_ds, vectors=english_fasttext_vectors, min_freq=10)

       # DataLoaderを作成します（torchtextの文脈では単純にiteraterと呼ばれています）
       train_dl = torchtext.legacy.data.Iterator(train_ds, batch_size=batch_size, train=True)
       val_dl   = torchtext.legacy.data.Iterator(val_ds, batch_size=batch_size, train=False, sort=False)
       test_dl  = torchtext.legacy.data.Iterator(test_ds, batch_size=batch_size, train=False, sort=False)

       return train_dl, val_dl, test_dl, TEXT
    else:
       ################################################################################
       # torchtext 0.16.0版向けデータローダ
       ################################################################################
       import torch
       import torchtext
       import torchdata.datapipes as dp
       from   torchtext.data.utils import get_tokenizer
       from   torchtext.vocab import vocab
       from   torchtext.vocab import build_vocab_from_iterator
       from   torchtext.data.functional import to_map_style_dataset
       ################################################################################
       # fasttextベクトル読み込みと追加ベクトル
       ################################################################################
       english_fasttext_vectors = Vectors(name='data/wiki-news-300d-1M.vec')
       # <unk>等ベクトル追加
       # 新しいトークン用の埋め込みベクトルを用意する
       # 新しいトークンの数と埋め込みの次元数
       num_new_tokens = 4
       embedding_dim = english_fasttext_vectors.vectors.shape[1]  # FastTextの埋め込み次元に合わせる

       # とりあえず0ベクトルを基に要所を1にして作成
       new_token_embeddings = torch.zeros(num_new_tokens, embedding_dim)
       new_token_embeddings[0,0]=1 #<unk>
       new_token_embeddings[1,1]=1 #<pad>
       new_token_embeddings[2,2]=1 #<bos>
       new_token_embeddings[3,3]=1 #<eos>
       # 既存の埋め込みベクトルをテンソルとして取得
       old_embeddings = english_fasttext_vectors.vectors
       # 既存の埋め込みベクトルに新しいトークンの埋め込みを追加
       updated_embeddings = torch.cat([new_token_embeddings,old_embeddings], dim=0)
       ################################################################################
       #単語辞書作成
       ################################################################################
       # 基の辞書english_fasttext_vectors.stoiに<unk>等ベクトル追加してvocabにする
       specials = ('<unk>', '<pad>', '<bos>', '<eos>')
       text_vocab = vocab(ordered_dict = english_fasttext_vectors.stoi, min_freq = 0, specials = specials, special_first = True)

       ################################################################################
       #本文読み込み,トークナイズ
       ################################################################################
       datapipe_trin = dp.iter.FileOpener(['./data/IMDb_train.tsv'], mode='rt', encoding="utf-8")
       datapipe_trin = dp.iter.CSVParser(datapipe_trin, delimiter='\t')
       datapipe_test = dp.iter.FileOpener(['./data/IMDb_test.tsv'], mode='rt', encoding="utf-8")
       datapipe_test = dp.iter.CSVParser(datapipe_test, delimiter='\t')
       # trainから valと分ける
       datapipe_trin, datapipe_val = datapipe_trin.random_split(total_length=len(list(datapipe_trin)), weights={"train": 0.8, "val": 0.2}, seed=42)

       # tokenizer設定
       tokenizer = get_tokenizer(tokenizer=tokenizer_with_preprocessing)
       #単語分割
       datapipe_trin = dp.iter.Mapper(datapipe_trin, lambda text: tokenizer(text), input_col=0)
       datapipe_val  = dp.iter.Mapper(datapipe_val,  lambda text: tokenizer(text), input_col=0)
       datapipe_test = dp.iter.Mapper(datapipe_test, lambda text: tokenizer(text), input_col=0)

       ################################################################################
       #  ミニバッチ作成の前処理
       ################################################################################
       import torchtext.transforms as T
       
       def loader_core(datapipe):
          datapipe_text = dp.iter.Mapper(datapipe, fn=lambda x: x[0])
          text_vocab = build_vocab_from_iterator(datapipe_text, specials=specials)
          text_vocab.get_itos()

          datapipe_label = dp.iter.Mapper(datapipe, fn=lambda x: x[1])
          label_vocab = build_vocab_from_iterator(datapipe_label)
          label_vocab.get_stoi()

          text_transform = T.Sequential(                           # 文の前処理
              T.VocabTransform(text_vocab),
              T.AddToken(token=text_vocab['<bos>'], begin=True),   # 文の開始前に<bos>をおく
              T.AddToken(token=text_vocab['<eos>'], begin=False),  # 文の末尾後に<eos>をおく
              T.Truncate(max_length),                              # トークン数をmax_length に制限する
              T.ToTensor(padding_value=text_vocab['<pad>']),       # テンソル化
              T.PadTransform(max_length,True)                      # うまく動作しない
          )

          label_transform = T.Sequential(                          # ラベルの前処理
              T.VocabTransform(label_vocab),
              T.ToTensor()
          )
          # ミニバッチ作成時のデータ変換関数
          def collate_batch(batch):
              texts  = text_transform([text for (text, label) in batch])
              labels = label_transform([label for (text, label) in batch])
              labels = torch.tensor([int(label_vocab.lookup_token(label)) for label in labels])  # 勝手に定義されたインデックス順{'1': 0, '0': 1}に対応させる
              return texts, labels

          ################################################################################
          # DataLoader設定 ミニバッチ作成
          ################################################################################
          from torch.utils.data import DataLoader
          from torchtext.data.functional import to_map_style_dataset
          ds = to_map_style_dataset(datapipe)
          data_loader = DataLoader(ds, batch_size=batch_size, shuffle=True, collate_fn=collate_batch)

          #for i, (texts, labels) in enumerate(data_loader):  # ばっちごとのループ  batch_size=5 で データ数=20なら 20/5=4回ループ
          #    print(i,len(texts),len(labels))
          #    for text, label in zip(texts, labels):
          #        print(text,label)
          #    #break
          #code.InteractiveConsole(globals()).interact()

          return data_loader

       data_loader_trin = loader_core(datapipe_trin)
       data_loader_val  = loader_core(datapipe_val)
       data_loader_test = loader_core(datapipe_test)
       return data_loader_trin,data_loader_val,data_loader_test,updated_embeddings
